tests/subsys/audio_module/src/bad_param_test.c

```diff
@@ -0,0 +1,1171 @@
+		      ret);
+
+	handle_tx.state = AUDIO_MODULE_STATE_UNDEFINED;
+	handle_rx.state = AUDIO_MODULE_STATE_CONFIGURED;
+	ret = audio_module_connect(&handle_tx, &handle_rx, false);
+	zassert_equal(ret, -ECANCELED, "Connect function did not return -ECANCELED (-129): ret %d",
+		      ret);
+}
+
+ZTEST(suite_audio_module_bad_param, test_connect_null)
+{
+	int ret;
+
+	test_initialise_description(&test_description_tx, "Module Test 1", AUDIO_MODULE_TYPE_IN_OUT,
+				    NULL);
+	test_initialise_description(&test_description_rx, "Module Test 2", AUDIO_MODULE_TYPE_IN_OUT,
+				    NULL);
+
+	test_initialise_handle(&handle_tx, "TEST Connect 1", &test_description_tx,
+			       AUDIO_MODULE_STATE_CONFIGURED, NULL, NULL);
+	test_initialise_handle(&handle_rx, "TEST Connect 2", &test_description_rx,
+			       AUDIO_MODULE_STATE_CONFIGURED, NULL, NULL);
+
+	ret = audio_module_connect(NULL, NULL, false);
+	zassert_equal(ret, -EINVAL, "Connect function did not return -EINVAL (%d): ret %d", -EINVAL,
+		      ret);
+
+	ret = audio_module_connect(NULL, &handle_rx, false);
+	zassert_equal(ret, -EINVAL, "Connect function did not return -EINVAL (%d): ret %d", -EINVAL,
+		      ret);
+
+	ret = audio_module_connect(&handle_tx, NULL, false);
+	zassert_equal(ret, -EINVAL, "Connect function did not return -EINVAL (%d): ret %d", -EINVAL,
+		      ret);
+
+	ret = audio_module_connect(&handle_tx, &handle_rx, false);
+	zassert_equal(ret, 0, "Connect function did not return 0L (0): ret %d", ret);
+
+	ret = audio_module_connect(&handle_tx, &handle_rx, false);
+	zassert_equal(ret, -EALREADY, "Connect function did not return -EALREADY (%d): ret %d",
+		      -EALREADY, ret);
+
+	ret = audio_module_connect(&handle_tx, NULL, true);
+	zassert_equal(ret, -EINVAL, "Connect function did not return -EALREADY (%d): ret %d",
+		      -EINVAL, ret);
+}
+
+ZTEST(suite_audio_module_bad_param, test_close_bad_state)
+{
+	int ret;
+
+	test_initialise_description(&test_description, "Module Test", AUDIO_MODULE_TYPE_IN_OUT,
+				    NULL);
+
+	test_initialise_handle(&handle, "TEST Close", &test_description,
+			       AUDIO_MODULE_STATE_CONFIGURED, NULL, NULL);
+
+	ret = audio_module_close(NULL);
+	zassert_equal(ret, -EINVAL, "Close function did not return -EALREADY (%d): ret %d", -EINVAL,
+		      ret);
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_CONFIGURED,
+		      "Close returns with incorrect state: %d", handle.state);
+
+	handle.state = AUDIO_MODULE_STATE_UNDEFINED;
+	ret = audio_module_close(&handle);
+	zassert_equal(ret, -ECANCELED, "Close function did not return -ECANCELED (%d): ret %d",
+		      -ECANCELED, ret);
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_UNDEFINED,
+		      "Close returns with incorrect state: %d", handle.state);
+
+	handle.state = AUDIO_MODULE_STATE_RUNNING;
+	ret = audio_module_close(&handle);
+	zassert_equal(ret, -ECANCELED, " did not return -ECANCELED (%d): ret %d", -ECANCELED, ret);
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_RUNNING,
```

Comment:
I don't get this test. The module is set to state RUNNING, we close it and then it should still be in state RUNNING?

Reply:
This is a test to ensure that a running module is not closed as this could have undetermined effects. The process is to stop the module and then close it. Hence the close should return -ECANCELED and remain in the AUDIO_MODULE_STATE_RUNNING state.